File: strings/1957-delete-characters-to-make-fancy-string.go
package strings

// https://leetcode.com/problems/delete-characters-to-make-fancy-string/

// Approach: Stack
// Time: O(n)
// Space: O(n)
func makeFancyString(s string) string {
	var stack = make([]rune, 0, len(s))
	var size = 0
	for _, c := range s {
		if size < 2 || stack[size-1] != c || stack[size-2] != c {
			stack = append(stack, c)
			size++
		}
	}
	return string(stack)
}

// // Approach: Two Pointers
// // Time: O(n)
// // Space: O(n)
// func makeFancyString(s string) string {
// 	var n = len(s)
// 	var fancy = []byte(s)
// 	var slow = 2
// 	for fast := 2; fast < n; fast++ {
// 		if fancy[fast] != fancy[slow-1] || fancy[fast] != fancy[slow-2] {
// 			fancy[slow] = s[fast]
// 			slow++
// 		}
// 	}
// 	return string(fancy[:min(slow, len(s))])
// }

File: strings/1957-delete-characters-to-make-fancy-string.py
import unittest

# https://leetcode.com/problems/delete-characters-to-make-fancy-string/

# python3 -m unittest strings/1957-delete-characters-to-make-fancy-string.py


class Solution(unittest.TestCase):
    # # Approach: Two Pointers
    # # Time: O(n)
    # # Space: O(n)
    # def makeFancyString(self, s: str) -> str:
    #     fancy = list(s)
    #     slow = 2
    #     for fast in range(2, len(s)):
    #         if s[fast] != fancy[slow-1] or s[fast] != fancy[slow-2]:
    #             fancy[slow] = s[fast]
    #             slow += 1
    #     return "".join(fancy[:min(len(s), slow)])

    # Approach: Stack
    # Time: O(n)
    # Space: O(n)
    def makeFancyString(self, s: str) -> str:
        stack = []
        for c in s:
            if len(stack) < 2 or stack[-1] != c or stack[-2] != c:
                stack.append(c)
        return "".join(stack)

    # def makeFancyString(self, s: str) -> str:
    #     fancy, freq = [s[0]], 1
    #     for i in range(1, len(s)):
    #         if s[i - 1] == s[i]:
    #             freq += 1
    #         else:
    #             freq = 1
    #         if freq < 3:
    #             fancy.append(s[i])
    #     return "".join(fancy)

    def test(self):
        for s, expected in [
            ("a", "a"),
            ("aab", "aab"),
            ("aaabaaaa", "aabaa"),
            ("leeetcode", "leetcode"),
            ("leeetcodeeeeeeeeeeeeeeeeeeeeeedddddddddddddddddddddd", "leetcodeedd"),
        ]:
            output = self.makeFancyString(s)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
